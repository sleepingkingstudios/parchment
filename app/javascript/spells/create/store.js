import submitForm from '../../requests/submitForm';
import { buildSpell } from '../entities';

const namespace = 'createSpell';
const REQUEST_URL = '/api/spells';

export const {
  actions,
  reducer,
  request,
} = submitForm({
  data: { spell: buildSpell() },
  namespace,
  url: REQUEST_URL,
});
