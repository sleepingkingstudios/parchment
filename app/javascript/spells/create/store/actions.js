const namespace = 'spells/create';

export const REQUEST_FAILURE = `${namespace}/requestFailure`;
export const REQUEST_PENDING = `${namespace}/requestPending`;
export const REQUEST_SUCCESS = `${namespace}/requestSuccess`;

export const requestFailure = errors => ({
  type: REQUEST_FAILURE,
  payload: { errors },
});

export const requestPending = () => ({
  type: REQUEST_PENDING,
  payload: {},
});

export const requestSuccess = data => ({
  type: REQUEST_SUCCESS,
  payload: { data },
});
