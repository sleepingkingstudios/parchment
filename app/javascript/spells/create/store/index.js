import submitForm from '../../../requests/submitForm';
import { buildSpell } from '../../entities';
import alerts from './alerts';
import redirect from './redirect';

const namespace = 'createSpell';
const REQUEST_URL = '/api/spells';

export const {
  actions,
  reducer,
  request,
} = submitForm({
  data: { spell: buildSpell() },
  middleware: [
    redirect,
    alerts,
  ],
  namespace,
  url: REQUEST_URL,
});
