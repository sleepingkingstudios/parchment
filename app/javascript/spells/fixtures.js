/* eslint-disable-next-line import/prefer-default-export */
export const spellsData = [
  {
    id: 'e17bca01-24d3-48d3-a308-099dd6dca373',
    level: 0,
    name: 'Embers',
    school: 'Evocation',
  },
  {
    id: 'e3641970-fc98-4674-b873-5532d89d9f26',
    level: 3,
    name: "Dragon's Breath",
    school: 'Evocation',
  },
  {
    id: 'a74d8c21-607d-4a3f-8cfd-2bbb48ec196a',
    level: 9,
    name: 'Wrath of the Titans',
    school: 'Evocation',
  },
];
