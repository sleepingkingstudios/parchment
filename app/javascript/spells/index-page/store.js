import buildStore from 'resource/pages/index-page/store';

const namespace = 'spells/index';
const resourceName = 'spells';
const url = 'api/spells';

const store = buildStore({
  namespace,
  resourceName,
  url,
});
const {
  hooks,
  reducer,
} = store;

export default store;

export {
  hooks,
  reducer,
};
