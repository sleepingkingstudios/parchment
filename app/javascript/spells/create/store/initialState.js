import { INITIALIZED } from '../../../store/requestStatus';

export default {
  errors: {},
  requestStatus: INITIALIZED,
};
