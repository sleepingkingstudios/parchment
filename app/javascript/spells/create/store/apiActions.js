import fetch from 'cross-fetch';
import { push } from 'connected-react-router';

import {
  requestFailure,
  requestPending,
  requestSuccess,
} from './actions';
import { addAlert } from '../../../components/alerts/store/actions';
import { formatErrors } from '../../../components/form/utils';
import {
  camelizeKeys,
  underscoreKeys,
} from '../../../utils/object';
import { generateFingerprintUuid } from '../../../utils/uuid';

export const SPELLS_API_URL = '/api/spells';

const handleRequestFailure = dispatch => async (response) => {
  const json = await response.json();
  const { errors } = json;

  dispatch(requestFailure(formatErrors(errors)));

  const alert = {
    id: generateFingerprintUuid('spells/create'),
    alertStyle: 'warning',
    dismissible: true,
    message: 'Unable to create spell.',
  };
  dispatch(addAlert(alert));
};

const handleRequestSuccess = dispatch => async (response) => {
  const json = await response.json();
  const { spell } = camelizeKeys(json.data);

  dispatch(requestSuccess(spell));

  const spellUrl = `/spells/${spell.id}`;
  dispatch(push(spellUrl));

  const alert = {
    id: generateFingerprintUuid('spells/create'),
    alertStyle: 'success',
    dismissible: true,
    message: 'Successfully created spell.',
  };
  dispatch(addAlert(alert));
};

const handleRequestResponse = dispatch => (response) => {
  if (response.ok) {
    handleRequestSuccess(dispatch)(response);
  } else {
    handleRequestFailure(dispatch)(response);
  }
};

export const performRequest = () => async (dispatch, getState) => {
  dispatch(requestPending());

  const state = getState();
  const { createSpell } = state;
  const data = underscoreKeys(createSpell.data);

  const url = SPELLS_API_URL;
  const response = await fetch(url, {
    method: 'POST',
    body: JSON.stringify(data),
    headers: {
      'Content-Type': 'application/json',
    },
  });

  return handleRequestResponse(dispatch)(response);
};
