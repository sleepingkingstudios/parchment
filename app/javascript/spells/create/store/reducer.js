import initialState from './initialState';
import {
  REQUEST_FAILURE,
  REQUEST_PENDING,
  REQUEST_SUCCESS,
} from './actions';
import {
  FAILURE,
  PENDING,
  SUCCESS,
} from '../../../store/requestStatus';

const createSpellReducer = (state = initialState, action) => {
  switch (action.type) {
    case REQUEST_FAILURE:
      /* eslint-disable-next-line no-console */
      console.log('createSpellReducer: REQUEST_FAILURE', action.payload);
      return Object.assign({}, state, {
        requestStatus: FAILURE,
      });
    case REQUEST_PENDING:
      /* eslint-disable-next-line no-console */
      console.log('createSpellReducer: REQUEST_PENDING');
      return Object.assign({}, state, {
        requestStatus: PENDING,
      });
    case REQUEST_SUCCESS:
      /* eslint-disable-next-line no-console */
      console.log('createSpellReducer: REQUEST_SUCCESS', action.payload);
      return Object.assign({}, state, {
        requestStatus: SUCCESS,
      });
    default:
      return state;
  }
};

export default createSpellReducer;
