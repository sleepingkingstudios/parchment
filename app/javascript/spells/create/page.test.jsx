import React from 'react';
import { shallow } from 'enzyme';

import CreateSpellPage from './page';

import { INITIALIZED } from '../../store/requestStatus';
import { spellsData } from '../fixtures';

describe('<CreateSpellPage />', () => {
  const performRequest = jest.fn();
  const updateFormField = jest.fn();
  const defaultProps = {
    data: spellsData[0],
    errors: { name: ["can't be blank"] },
    performRequest,
    requestStatus: INITIALIZED,
    updateFormField,
  };
  const breadcrumbs = [
    {
      label: 'Home',
      url: '/',
    },
    {
      label: 'Spells',
      url: '/spells',
    },
    {
      label: 'Create',
      url: '/spells/create',
      active: true,
    },
  ];

  it('should wrap the contents in a Page', () => {
    const rendered = shallow(<CreateSpellPage {...defaultProps} />);

    expect(rendered).toHaveDisplayName('Page');
    expect(rendered).toHaveClassName('page-spells');
    expect(rendered).toHaveProp('breadcrumbs', breadcrumbs);
  });

  it('should render the spell form', () => {
    const rendered = shallow(<CreateSpellPage {...defaultProps} />);
    const form = rendered.find('SpellForm');
    const {
      data,
      errors,
      requestStatus,
    } = defaultProps;

    expect(form).toExist();
    expect(form).toHaveProp('onChangeAction', updateFormField);
    expect(form).toHaveProp('onSubmitAction', performRequest);
    expect(form).toHaveProp('data', data);
    expect(form).toHaveProp('errors', errors);
    expect(form).toHaveProp('requestStatus', requestStatus);
  });

  it('should match the snapshot', () => {
    const rendered = shallow(<CreateSpellPage {...defaultProps} />);

    expect(rendered).toMatchSnapshot();
  });
});
