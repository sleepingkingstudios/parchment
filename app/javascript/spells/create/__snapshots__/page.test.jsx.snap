// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateSpellPage /> should match the snapshot 1`] = `
<Page
  breadcrumbs={
    Array [
      Object {
        "label": "Home",
        "url": "/",
      },
      Object {
        "label": "Spells",
        "url": "/spells",
      },
      Object {
        "active": true,
        "label": "Create",
        "url": "/spells/create",
      },
    ]
  }
  className="page-spells page-create-spell"
  layout={null}
  subtitle="5e Campaign Companion"
  title="Parchment"
>
  <h1>
    Create Spell
  </h1>
  <SpellForm
    data={
      Object {
        "castingTime": "1 reaction, which you take when you speak your own name or hear it spoken",
        "description": "Immediately after speaking or hearing your name, you create the illusory sound of an ominous, distant thunderclap.",
        "duration": "Instantaneous",
        "id": "e17bca01-24d3-48d3-a308-099dd6dca373",
        "level": 0,
        "materialComponent": "",
        "name": "Thunderclap",
        "range": "Self",
        "ritual": false,
        "school": "illusion",
        "shortDescription": "Create an ominous thunderclap when your name is spoken",
        "somaticComponent": false,
        "verbalComponent": false,
      }
    }
    errors={
      Object {
        "name": Array [
          "can't be blank",
        ],
      }
    }
    isUpdate={false}
    onChangeAction={[MockFunction]}
    onSubmitAction={[MockFunction]}
    requestStatus="initialized"
  />
</Page>
`;
